### PR TITLE
Integration of Victron SmartShunt via VE.Direct

### PR DESCRIPTION
**Pull request by philippsandhaus**
This a first draft of my integration as discussed in https://github.com/helgeerbe/OpenDTU-OnBattery/issues/205. As I did quite a lot of restructuring and I am not that confident in C++ I am marking this as draft and hereby ask for feedback from @schlimmchen and/or @helgeerbe. I rebased the current status on top of the development branch.

Bildschirmfoto 2023-09-17 um 22 38 20
Bildschirmfoto 2023-09-17 um 22 08 23

### CODE REVIEW

**Comment by helgeerbe** [comment]
@philippsandhaus good job! I just looked at the frame handler. I would propose to use a `VeDirectController`base class and derive 'veDirectShuntController`and `VeDirectMpptController`from it. 

Advantage is, that you can do the generic stuff like init of the HardwareSerial in the base class and you have to implement it only once. Do only device specific in the DeviceController class.

File: webapp/src/locales/en.json [hunk 1]
```diff
@@ -608,6 +608,7 @@
         "Provider": "Data Provider",
         "ProviderPylontechCan": "Pylontech using CAN bus",
         "ProviderJkBmsSerial": "Jikong (JK) BMS using serial connection",
+        "ProviderVictron": "Victron SmartShunt using VE.Direct Interface",
```
Comment (schlimmchen) [inline webapp/src/locales/en.json:611]: This is missing in fr.json.

**Review by schlimmchen — APPROVED** [review]

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.h [hunk 2]
```diff
@@ -66,6 +52,9 @@ class MovingAverage {
         return static_cast<double>(_sum) / _count;
     }
 
+
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectFrameHandler.h:55]: Noting ;-) I removed them.

File: include/BatteryStats.h [hunk 3]
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
...
 };
+
+class VictronSmartShuntStats : public BatteryStats {
+    public:
+        void getLiveViewData(JsonVariant& root) const final;
+        void mqttPublish() const final;
+
+        void updateFrom(VeDirectShuntController::veShuntStruct const& shuntData);
+
+        void setSoC(uint8_t SoC) { _SoC = SoC; _lastUpdateSoC = millis(); }
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply include/BatteryStats.h:110]: Yes, removed it

File: include/BatteryStats.h [hunk 4]
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
...
+        void setLastUpdate(uint32_t ts) { _lastUpdate = ts; }
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply include/BatteryStats.h:111]: Yes, removed it.

File: include/BatteryStats.h [hunk 5]
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
...
+
+        float _voltage;
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply include/BatteryStats.h:113]: Yes, I did. Made them private now.

File: lib/VeDirectFrameHandler/VeDirectMpptController.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+{
+}
+
+void VeDirectMpptController::init(int8_t rx, int8_t tx, Print* msgOut, bool verboseLogging)
+{
+    _vedirectSerial = new HardwareSerial(1);
+	_vedirectSerial->begin(19200, SERIAL_8N1, rx, tx);
+    _vedirectSerial->flush();
+	VeDirectFrameHandler::init(msgOut, verboseLogging);
+	if (_verboseLogging) _msgOut->println("Finished init MPPTCOntroller");
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectMpptController.cpp:16]: Done

File: webapp/src/locales/de.json [hunk 7]
```diff
@@ -599,6 +599,7 @@
         "Provider": "Datenanbieter",
         "ProviderPylontechCan": "Pylontech per CAN-Bus",
         "ProviderJkBmsSerial": "Jikong (JK) BMS per serieller Verbindung",
+        "ProviderVictron": "Victron SmartShunt per VE.Direct Interface",
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply webapp/src/locales/de.json:602]: Done

File: webapp/src/locales/en.json (hunk 1 above)
Comment (philippsandhaus) [reply webapp/src/locales/en.json:611]: Hmm, the whole batteryadmin section is missing for. I copied the complete section from en for now. I guess the whole file has to be translated at some point.

File: src/VictronSmartShunt.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    _verboseLogging = verboseLogging;
+
+    MessageOutput.println(F("[VictronSmartShunt] Initialize interface..."));
+
+    const PinMapping_t& pin = PinMapping.get();
+    MessageOutput.printf("[VictronSmartShunt] Interface rx = %d, tx = %d\r\n",
+            pin.battery_rx, pin.battery_tx);
+
+    if (pin.battery_rx < 0 || pin.battery_tx < 0) {
+        MessageOutput.println(F("[Pylontech] Invalid pin config"));
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply src/VictronSmartShunt.cpp:19]: Adjusted ;-)

File: webapp/src/locales/en.json [hunk 9]
```diff
@@ -835,6 +836,10 @@
         "underVoltage": "Undervoltage",
         "highVoltage": "High voltage",
         "overVoltage": "Overvoltage",
-        "bmsInternal": "BMS internal"
+        "bmsInternal": "BMS internal",
+        "nrChargedCycles": "Number of charged cycles",
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply webapp/src/locales/en.json:840]: I just looked into the Victron App. It just calls it charge cycles (Ladezyklen). So I guess I just copy that.

File: webapp/src/locales/fr.json [hunk 10]
```diff
@@ -772,6 +772,10 @@
         "underVoltage": "Undervoltage",
         "highVoltage": "High voltage",
         "overVoltage": "Overvoltage",
-        "bmsInternal": "BMS internal"
+        "bmsInternal": "BMS internal",
+        "nrChargedCycles": "Number of charged cycles",
+        "chargedEnergy": "Charged energy",
+        "dischargedEnergy": "Discharged energy",
+        "model": "Model"
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply webapp/src/locales/fr.json:779]: It was in an intermediate version. I removed it now.

File: lib/VeDirectFrameHandler/VeDirectMpptController.h [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        int32_t H21;                    // maximum power today W
+        double H22;                     // yield yesterday kWh
+        int32_t H23;                    // maximum power yesterday W
+    };
+
+    veMpptStruct veFrame{}; 
+
+private:
+
+    virtual void textRxEvent(char * name, char * value);
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectMpptController.h:38]: Thanks, this is exactly the kind of feedback I need. Forces me to understand the meanings of final, virtual, override etc. I guess final is the right one here.

File: lib/VeDirectFrameHandler/VeDirectShuntController.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        double H18;                     // Amount of charged energy
+    };
+
+    veShuntStruct veFrame{}; 
+
+    
+private:
+
+    virtual void textRxEvent(char * name, char * value);
+    void frameEndEvent(bool) override;                 // copy temp struct to public struct
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectShuntController.h:48]: Changed to final

File: lib/VeDirectFrameHandler/VeDirectShuntController.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        double H17;                     // Amount of discharged energy
+        double H18;                     // Amount of charged energy
+    };
+
+    veShuntStruct veFrame{}; 
+
+    
+private:
+
+    virtual void textRxEvent(char * name, char * value);
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectShuntController.h:47]: Changed to final

**Comment by philippsandhaus** [comment]
Thanks for the review. I adjusted my code accordingly and also moved initialization of the HardwareSerial interface to the base class. As far as I know there are only 2 HardwareSerial interfaces on the ESP32, right? So if there are more than 2 serial devices, one would have to think about some kind of management and also use SoftwareSerial, I guess.

Yes, for me the MPPT interface still works as before.

File: src/VictronSmartShunt.cpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    if (pin.battery_rx < 0 || pin.battery_tx < 0) {
+        MessageOutput.println(F("[VictronSmartShunt] Invalid pin config"));
+        return false;
+    }
+
+    auto tx = static_cast<gpio_num_t>(pin.battery_tx);
+    auto rx = static_cast<gpio_num_t>(pin.battery_rx);
+
+    VeDirectShunt.init(rx, tx, &MessageOutput, verboseLogging);
```
Comment (schlimmchen) [inline src/VictronSmartShunt.cpp:26]: `_veDirectShunt = std::make_unique<VeDirectShuntController>()` goes here. consider removing the VeDirectShuntController::init() method and doing its work in the constructor instead.

**Review by schlimmchen — CHANGES_REQUESTED** [review]

**Comment by helgeerbe** [comment]
I wouldn't mind, if we switch to Software Serial completely for ve.direct devices. Guess this makes the orchestration easier. @schlimmchen pointed out baudrate is moderate and the actual loop time is fast enough. 

Just thinking loud, if we are going to support also a second charger.

**Comment by philippsandhaus** [comment]
`SoftwareSerial` and `HardwareSerial` both derive from `Stream`. I envision a class like `SerialProvider` which provides instances of `Stream` with for example `SerialProvider.getSerial(19200, SERIAL_8N1, rx, tx)`. 'SerialProvider' could then make the decision to hand an instance of `HardwareSerial` or `SoftwareSerial` based on baudrate and/or availability of hardware serial ports. But I guess the tricky parts are the details like how to get notified when a Hardware Serial port is freed up.

**Comment by schlimmchen** [comment]
@philippsandhaus Interesting thoughts! How about you open up a new issue where we can discuss this (rather than being buried in this PR)?

**Comment by schlimmchen** [comment]
@philippsandhaus I see there are values in `veShuntStruct` that are of type double. However, all values are always converted using `atoi`. I think it's fine to save the values as they were on the serial line, i.e., integers, and to do the scaling later -- even though this is different than in the MPPT counterpart. You do the scaling in `VictronSmartShuntStats::updateFrom`. If so, shouldn't the values in the `veShuntStruct` be all `int32_t`?

**Comment by schlimmchen** [comment]
@philippsandhaus Feel free to cherry pick from https://github.com/schlimmchen/OpenDTU-OnBattery/commits/devel for this PR -- if you like. And except for the SoftwareSerial commit, that's a work in progress and requires testing.

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp [hunk 15]
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
 	if (!_verboseLogging) { _debugIn = 0; }
 }
 
-void VeDirectFrameHandler::init(int8_t rx, int8_t tx, Print* msgOut, bool verboseLogging)
+void VeDirectFrameHandler::init(Print* msgOut, bool verboseLogging, uint16_t rx, uint16_t tx, uint16_t hwSerialPort)
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp]: Cherry picked your changes

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp [hunk 16]
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
...
 {
-	VedirectSerial.begin(19200, SERIAL_8N1, rx, tx);
-	VedirectSerial.flush();
+	_vedirectSerial = new HardwareSerial(hwSerialPort);
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp]: Cherry picked your changes

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp [hunk 17]
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
...
+	_vedirectSerial->begin(19200, SERIAL_8N1, rx, tx);
+    _vedirectSerial->flush();
```
[the remark this answers is not in the document]
Comment (philippsandhaus) [reply lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp]: Cherry picked your changes

**Comment by philippsandhaus** [comment]
Great, I have cherry picked most of your adjustments. I also added handling of alarm cases for the live view. At least for me this PR would now be complete.

**Comment by schlimmchen** [comment]
> At least for me this PR would now be complete.

Affirmative.

**Comment by helgeerbe** [comment]
Cool. I could merge it into development. But I'm away from keyboard next week. So I would keep it in development for one additional week for further testing, before I release it. Are you fine with that?

**Comment by swingstate** [comment]
I'll install a SmartShunt 500A tomorrow morning and was assuming you're going to release, however no worries - I'll compile development and flash that to my DTU for testing. Will report any findings and if no issue report back a "no issues"

**Comment by philippsandhaus** [comment]
> Cool. I could merge it into development. But I'm away from keyboard next week. So I would keep it in development for one additional week for further testing, before I release it. Are you fine with that?

Sure, no problem.

**Comment by swingstate** [comment]
guess now it is. I'll compile and flash my ESP tomorrow morning.

**Comment by helgeerbe** [comment]
It's in development now. And it's on my OpenDTU.
Happy testing!

**Comment by helgeerbe** [comment]
I removed the line. From my point of view. You should do the build locally. 
Action is only used to ensure, that every commit to development results in code that is compilable.

**Comment by swingstate** [comment]
@helgeerbe I have the code on my ESP now and so far things look good, SOC is being displayed. I have one unclarity which I shared in the longer issue thread #205.

**Pull request merged**